### PR TITLE
Update core bindings to Yrs v0.18

### LINKED ISSUE #91: Update core bindings to Yrs v0.18
With the release of [Yrs v0.18](https://github.com/y-crdt/y-crdt/releases/tag/v0.18), some core functions changed and should be updated in the library.

- [x] Replace `*_unobserve_*()` bindings with call to `yunobserve()`
  - Refers to: New Observer API [#385@y-crdt](https://github.com/y-crdt/y-crdt/pull/385)
- [x] Remove `yxmlelem()`
  - [x] Fix unit tests
- [x] Remove `yxmltext()`
  - [x] Fix unit tests
- [x] Remove `ybranch_write_transaction()`
- [x] Remove `ybranch_read_transaction()`
- [x] Create bindings for `YBranchId`:
  - [x] `ybranch_id()`
  - [x] `ybranch_get()`
  - [x] `ybranch_alive()`
  - Refers to: Logical pointers [#393@y-crdt](https://github.com/y-crdt/y-crdt/pull/393)

All items above include the bindings and unit tests.

---

Breaking changes
- Remove `Doc.XmlElement()`
  - Fix: use `Doc.XmlFragment()` for root-level types and then create an `XmlElement` inside of the `XmlFragment`.
- Remove `Doc.XmlText()`
  - Fix: use `Doc.XmlFragment()` for root-level types and then create an `XmlText` inside of the `XmlFragment`.

### PR DESCRIPTION
**Pull request by LSViana**
Closes #91.

This PR is a WIP. There are some `TODO` comments in the code to be checked.

### CODE REVIEW

File: YDotNet/Native/Types/Branches/BranchKind.cs
Comment (SebastianStehle) [inline YDotNet/Native/Types/Branches/BranchKind.cs:7]: I don't like this, because it creates a circular dependency between Native namespace and the higher level Document  namespace. 

We defined, that Document depends on Native, not the other direction.

File: YDotNet/Document/Transactions/Transaction.cs
Comment (SebastianStehle) [inline YDotNet/Document/Transactions/Transaction.cs:123]: I think we should add this spacing setting to editorconfig or something like that, so that we don't override each other settings

**Review by SebastianStehle — CHANGES_REQUESTED** [review]
IMHO we should discuss the is_alive feature. I have discussed this several times with the team and it is actually a solution to prevent unexpected crashes in .NET because of dangling pointers.

What we should do, is to call this method before each access to the underlaying pointer to guard our library and to prevent server crashes that cannot be logged or catched.

**Comment by LSViana** [comment]
> IMHO we should discuss the is_alive feature. I have discussed this several times with the team and it is actually a solution to prevent unexpected crashes in .NET because of dangling pointers.
> 
> What we should do, is to call this method before each access to the underlaying pointer to guard our library and to prevent server crashes that cannot be logged or catched.

I agree. Would it work if we overrode `ThrowIfDisposed()` in the `Branch` class to also check for `Alive()`?

This would allow us to run this check without changing a lot of code since the `Branch` and all subclasses already use `ThrowIfDisposed()` before calling all methods (I didn't find any method lacking this check, for example).

File: YDotNet/Native/Types/Branches/BranchKind.cs
Comment (LSViana) [reply YDotNet/Native/Types/Branches/BranchKind.cs]: That makes sense, I actually missed this detail when I was applying this change. My goal here was to reduce the duplication of these constants in the project.

But, after all, I guess we'll have to keep them. I just pushed [this commit](https://github.com/y-crdt/ydotnet/commit/94dc35c91635d3b8f0415a1e6dc4cdb051829064) to revert this change.

File: YDotNet/Document/Transactions/Transaction.cs
Comment (LSViana) [reply YDotNet/Document/Transactions/Transaction.cs]: I just realized the StyleCop support was disabled in my IDE. 🫠

I enabled it and the IDE auto-formatted the file [like this](https://github.com/y-crdt/ydotnet/commit/76cfddc0e621bc0adf83b924028e0450fdf00258). Do you see the same formatting being applied on your side too?

**Comment by LSViana** [comment]
Hi, @SebastianStehle!

Did you have a chance to continue reviewing this PR? I assume you're not finished since there are just a couple of comments and I pushed a lot of changes. If you just didn't have time yet, that's fine too.

I'll probably be away from the computer for the next few days because of a local holiday but I might resume working on the changes during the weekend.

File: YDotNet/Document/Transactions/Transaction.cs
Comment (SebastianStehle) [reply YDotNet/Document/Transactions/Transaction.cs]: Yes, that should work.

**Comment by SebastianStehle** [comment]
Yes, I am. I have not reviewed the tests and everything else was looking fine :)

File: YDotNet/Native/Types/Branches/BranchKind.cs
Comment (LSViana) [reply YDotNet/Native/Types/Branches/BranchKind.cs]: @SebastianStehle I'm resolving this because this problem is now gone.

File: YDotNet/Document/Transactions/Transaction.cs
Comment (LSViana) [reply YDotNet/Document/Transactions/Transaction.cs]: Great! I'm resolving the discussion, then.

**Comment by LSViana** [comment]
Regarding this:

> I would ...
> 
> 1. Create a struct for ybranch_id, so we can distinguish it from actual pointers. Mark this internal.
> 2. Always store brand ID, never any pointers.
> 3. Provide a method to get the pointer within a transaction and use this in all methods. We could either call the internal alive method in the Disposed check method or just within this method and get rid of the general disposed method.
> 
> But in general **the branch_id is something internal and we should never store pointers.**

For item (1), I created the `internal struct BranchIdNative` so this part is done already. The change here is to remove the `public class BranchId` and handle it internally only. Is that correct?

For item (2), I'm already storing an instance of the `struct BranchIdNative` so this part shouldn't change.

For item (3), I'm not sure what you meant. Could you please explain your idea again and maybe the steps you'd use to implement it?

**Comment by SebastianStehle** [comment]
3) I just meant that we should not store pointers, only IDs. Only use pointers within a method as a temporary variable.

**Comment by LSViana** [comment]
Okay, this makes sense but I don't see how to reflect this in the code properly.

If you have the time, please feel free to push commits here to show what you mean, it's probably faster and more efficient.

**Comment by SebastianStehle** [comment]
I have pushed a test commit.

**Comment by SebastianStehle** [comment]
Perhaps we should focus on the breaking changes first without the ID stuff and then add them in a later reelase.

**Comment by LSViana** [comment]
@SebastianStehle I'm not sure I understand the latest changes. If we turn the `public class BranchId` into `internal record struct`, we can probably just drop it and use the `BranchIdNative` directly, no?

Besides that, most of the properties of the former `public class BranchId` aren't used anymore so we can drop them.

Maybe I'm misunderstanding the purpose of the branch IDs.

**Comment by SebastianStehle** [comment]
I just introduced the class to have some shared behavior like dispose handling.

BranchId is like a virtual pointer or reference to solve the problem of dangling pointers and reported crashes. Therefore there is no need to expose that, it is not meant to be used outside.

**Comment by LSViana** [comment]
@SebastianStehle Got it!

I understand it better now. I pushed three commits now that I'd like you to review.

If you agree with the changes, we can propagate them to all other shared types that inherit from the `Branch` class.

**Comment by SebastianStehle** [comment]
I have not seen it, but looks good to me.

**Comment by LSViana** [comment]
@SebastianStehle That's great! Given that, I propagated those changes to all other `Branch` types.

Can you check these latest changes in the classes of the `YDotNet` project, please?

**Comment by LSViana** [comment]
I just pushed some extra commits to fix all unit tests (including some API changes to `UndoEvent` to mirror changes from `yffi`).

The macOS tests are failing for some reason but I can check it later.

**Comment by LSViana** [comment]
@SebastianStehle If you have time, can you please check the latest changes mentioned in my previous comment?

File: YDotNet/Document/Events/EventSubscriber.cs
Comment (SebastianStehle) [inline YDotNet/Document/Events/EventSubscriber.cs:5]: Do we still need this file? I mean it is potentially unsafe to use the raw pointer.

File: YDotNet/Document/Types/Branches/Branch.cs
Comment (SebastianStehle) [inline YDotNet/Document/Types/Branches/Branch.cs:23]: Perhaps we can just rely on default here and avoid the nullabilty?

File: YDotNet/Document/Types/Branches/Branch.cs
Comment (SebastianStehle) [inline YDotNet/Document/Types/Branches/Branch.cs:44]: I just made a quick search, but I think this could be a private field.

File: YDotNet/Document/Types/Branches/Branch.cs
Comment (SebastianStehle) [inline YDotNet/Document/Types/Branches/Branch.cs:90]: Do we still need the disposed field? Can we not just use the branchId?

File: YDotNet/Document/Types/Branches/Branch.cs
Comment (SebastianStehle) [inline YDotNet/Document/Types/Branches/Branch.cs:95]: Using pattern would clarify the scope of the handle.

File: YDotNet/Document/Types/Events/EventBranch.cs
Comment (SebastianStehle) [inline YDotNet/Document/Types/Events/EventBranch.cs:72]: Do you remember why we do not use the type cache here? We should probably always use the cache or not at all.

**Review by SebastianStehle — COMMENTED** [review]
Only some minor things.

File: YDotNet/Document/Events/EventSubscriber.cs
Comment (LSViana) [reply YDotNet/Document/Events/EventSubscriber.cs:5]: No, we can't drop it because `Doc` and `UndoManager` aren't subclasses of `Branch` and still need this type of event manager.

File: YDotNet/Document/Types/Branches/Branch.cs
Comment (LSViana) [reply YDotNet/Document/Types/Branches/Branch.cs]: @SebastianStehle Is [this](https://github.com/y-crdt/ydotnet/commit/9d88a194d9dc7d8231baff7d9b379bba84fa6309) what you mean?

File: YDotNet/Document/Types/Branches/Branch.cs
Comment (LSViana) [reply YDotNet/Document/Types/Branches/Branch.cs]: @SebastianStehle Sure! Fixed by [this commit](https://github.com/y-crdt/ydotnet/commit/9298507a169665c9d86bf03f84fc83ef4abaee80).

File: YDotNet/Document/Types/Events/EventBranch.cs
Comment (LSViana) [reply YDotNet/Document/Types/Events/EventBranch.cs:72]: Not sure, IIRC this method was originally by you previously.

File: YDotNet/Document/Types/Branches/Branch.cs
Comment (LSViana) [reply YDotNet/Document/Types/Branches/Branch.cs:90]: @SebastianStehle I guess we do for the same reason mentioned in [this comment thread](https://github.com/y-crdt/ydotnet/pull/94#discussion_r1597751618).

Despite the changes to the subclasses of `Branch`, the `Doc` and `Manager` classes still need this because they're used in many places of `Doc`, for example, to validate the life cycle state.

**Comment by LSViana** [comment]
@SebastianStehle I checked and replied to your latest comments.

I'm not sure I fully understood all of your questions so, please, feel free to push code changes for details that you see are obvious to improve.

File: YDotNet/Document/Types/Branches/Branch.cs
Comment (SebastianStehle) [reply YDotNet/Document/Types/Branches/Branch.cs]: Yes

File: YDotNet/Document/Types/Branches/Branch.cs
Comment (SebastianStehle) [reply YDotNet/Document/Types/Branches/Branch.cs]: In these cases I usually use the normal braces, because it is only needed for the next line and not the whole method, but it does not really matter.

File: YDotNet/Document/Types/Events/EventBranch.cs
Comment (SebastianStehle) [reply YDotNet/Document/Types/Events/EventBranch.cs:72]: Ok, weird :D

**Comment by LSViana** [comment]
@SebastianStehle I just resolved all the latest comments that were pending. I'd just like to confirm with you if everything here is good to be merged and released, what do you think?

I'll also fix the failing macOS tests before merging the PR.

**Comment by SebastianStehle** [comment]
Yes, it is. I would like to know whether it is still needed to keep a reference to the document everywhere. But we can solve this after the PR.

**Comment by LSViana** [comment]
@SebastianStehle I just realized a few unit tests have been failing because the GitHub Action runner images use .NET 8 by default.

We have two options from here:
1. Manually configure them to use .NET 7 ([like this](https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-net#specifying-a-net-version))
2. Update the projects of the solution to .NET 8

Bumping a whole .NET version looks like a major change, then I'm leaning towards option #1 but option #2 is something we'll need to do at some point anyway.

What do you think?

**Comment by LSViana** [comment]
@SebastianStehle Also, can I get your approval for the PR? It's still marked as "requested changes". 🙂

**Comment by LSViana** [comment]
I'm going to merge this PR regardless of the pending approval.

**Pull request merged**
